NodeProg/Infrastructure/ParseHelpers.cs

```diff
@@ -24,77 +25,79 @@ public static class ParseHelpers
             if (input.Length < startIndex + length)
                 throw new ArgumentException("input.Length < startIndex + length", nameof(length));
 
-            if (length == 0)
-                return null;
-
-            var position = startIndex;
-            var current = input[position];
-            var endIndex = startIndex + length;
+            return TryParseInt64(inputEnumerable(input, startIndex, length), allowLeadingWhitespace, allowTrailingWhitespace);
+        }
+        public static long? TryParseInt64(
+            IEnumerable<char> input,
+            bool allowLeadingWhitespace = true,
+            bool allowTrailingWhitespace = true)
+        {
+            if (input == null)
+                throw new ArgumentNullException(nameof(input));
 
-            if (allowLeadingWhitespace)
+            using (var iChar = input.GetEnumerator())
             {
-                while (position < endIndex)
-                {
-                    current = input[position];
-                    if (current != ' ' && current != '\t')
-                        break;
-                    position++;
-                }
-                if (position == endIndex)
+                if (!iChar.MoveNext())
                     return null;
-                current = input[position];
-            }
-
-            var minus = 1;
 
-            if (current == '+' || current == '-')
-            {
-                if (current == '-')
-                    minus = -1;
-                position++;
-                if (position == endIndex)
-                    return null;
-                current = input[position];
-            }
+                if (allowLeadingWhitespace)
+                    while (iChar.Current == ' ' || iChar.Current == '\t')
+                        if (!iChar.MoveNext())
+                            return null;
 
-            long value = 0;
+                var minus = 1;
+                if (iChar.Current == '+' || iChar.Current == '-')
+                {
+                    if (iChar.Current == '-')
+                        minus = -1;
+                    if (!iChar.MoveNext())
+                        return null;
+                }
 
-            while (position < endIndex)
-            {
-                current = input[position];
+                long value = 0;
+                var iCharMN = true;
+                do
+                {
+                    if (iChar.Current < '0' || iChar.Current > '9')
+                        break;
+                    try
+                    {
+                        checked
+                        {
+                            value = value * 10 + minus * (iChar.Current - '0');
+                        }
+                    }
+                    catch
```

Comment:
use an exception type, be specific in what you expect to be thrown in case of numeric overflow.